cypress-io/cypress
# chore: implement experimental ESM stub/spy for Vite

- Closes https://github.com/cypress-io/cypress-internal/issues/564

### Additional details


First release of our Vite plugin to facilitate spy/mock on ES modules.

I made this `chore` since

1) It's still a POC and has known issues
2) I want it to say below v1 until we've tested it more with some of the pilot users

Not sure the best way to handle this, but `chore` should only increment the patch number.

### Steps to test


This is based on the PROPOSAL section from the tech brief: https://github.com/cypress-io/engineering-documentation/tree/main/tech-briefs/ct-esm-mocks#proposal. That explains what's going on here in detail.

It's complicated, don't feel you like need 100% understand it all when you review it. 

Testing: Run the test suite! Or make a new Vite project that uses ES Modules and write some tests using `cy.spy` and `cy.stub`. The things in our documentation for [`cy.stub()`](https://docs.cypress.io/api/commands/stub) and [`cy.spy()`](https://docs.cypress.io/api/commands/spy) should work correctly with ES modules. There's a bunch of tests in this PR demonstrating it works (mostly, sans the caveats in the README).

There are some caveats, I wrote those in the README.

### How has the user experience changed?


You can spy/mock ES modules (both static and dynamic imports).

### PR Tasks


- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
